tree/treeplayer/inc/TTreeReaderArray.h

```diff
@@ -59,90 +59,159 @@ namespace Internal {
-         // iterators to be equal.
-         if (!IsValid() && !lhs.IsValid())
+      bool operator==(const Iterator_t &other) const
+      {
+         // Follow C++14 requiring two past-the-end iterators to be equal.
+         if (!IsValid() && !other.IsValid())
             return true;
-         return fIndex == lhs.fIndex && fArray == lhs.fArray;
+         return fArray == other.fArray && fIndex == other.fIndex;
       }
 
-      bool operator!=(const Iterator_t& lhs) const {
-         // Compare not equal.
-         return !(*this == lhs);
+      bool operator!=(const Iterator_t &other) const { return !(*this == other); }
+
+      /// Pre-increment operator
+      Iterator_t &operator++()
+      {
+         ++fIndex;
+         if (fIndex >= fSize)
+            fArray = nullptr; // invalidate iterator
+         return *this;
       }
 
-      Iterator_t operator++(int) {
-         // Post-increment (it++).
-         Iterator_t ret = *this;
-         this->operator++();
+      /// Post-increment operator
+      Iterator_t operator++(int)
+      {
+         auto ret = *this;
+         auto &self = *this;
+         ++self;
          return ret;
       }
 
-      Iterator_t& operator++() {
-         // Pre-increment (++it).
-         if (IsValid()) {
-            ++fIndex;
-            if (fIndex >= fArray->GetSize()) {
-               // Remember that it's past-end.
-               fArray = 0;
-            }
-         }
+      /// Pre-decrement operator
+      Iterator_t &operator--()
+      {
+         if (fIndex == 0u)
+            fArray = nullptr; // invalidate iterator
+         else
+            --fIndex;
          return *this;
       }
 
-      T& operator*() const {
-         // Get the referenced element.
-         R__ASSERT(fArray && "invalid iterator!");
-         return fArray->At(fIndex);
+      /// Post-decrement operator
+      Iterator_t operator--(int)
+      {
+         auto ret = *this;
+         auto &self = *this;
+         --self;
+         return ret;
       }
 
-      operator const T*() const { return &fArray->At(fIndex); }
-   };
+      Iterator_t operator+(std::ptrdiff_t n) const { return Iterator_t(fIndex + n, fArray); }
+
+      Iterator_t operator-(std::ptrdiff_t n) const
+      {
+         const auto index = std::ptrdiff_t(fIndex);
+         const auto newIndex = index >= n ? index - n : std::numeric_limits<decltype(fIndex)>::max();
+         return Iterator_t(newIndex, fArray);
+      }
+
+      std::ptrdiff_t operator-(const Iterator_t &other) const { return fIndex - other.fIndex; }
+
+      Iterator_t &operator+=(std::ptrdiff_t n)
+      {
+         auto &self = *this;
+         self = self + n;
+         return self;
+      }
+
+      Iterator_t &operator-=(std::ptrdiff_t n)
+      {
+         auto &self = *this;
+         self = self - n;
+         return self;
+      }
 
-   typedef Iterator_t iterator;
+      bool operator<(const Iterator_t &other) const { return fIndex < other.fIndex; }
```

Comment:
Does `end() < end()` behave as expected?

Reply:
it should return `false` since both `fIndex` and `other.fIndex` are equal to `fSize`. Unless you meant something more subtle..?